src/PrimordialChem/test_primordial_chem.cpp

```diff
@@ -0,0 +1,239 @@
+//==============================================================================
+// TwoMomentRad - a radiation transport library for patch-based AMR codes
+// Copyright 2020 Benjamin Wibking.
+// Released under the MIT license. See LICENSE file included in the GitHub repo.
+//==============================================================================
+/// \file test_primordial_chem.cpp
+/// \brief Defines a test problem for primordial chemistry (microphysics).
+///
+#include <random>
+#include <vector>
+
+#include "AMReX_BC_TYPES.H"
+#include "AMReX_BLProfiler.H"
+#include "AMReX_BLassert.H"
+#include "AMReX_FabArray.H"
+#include "AMReX_GpuDevice.H"
+#include "AMReX_MultiFab.H"
+#include "AMReX_ParallelContext.H"
+#include "AMReX_ParallelDescriptor.H"
+#include "AMReX_ParmParse.H"
+#include "AMReX_SPACE.H"
+#include "AMReX_TableData.H"
+
+#include "RadhydroSimulation.hpp"
+#include "hydro_system.hpp"
+#include "radiation_system.hpp"
+#include "test_primordial_chem.hpp"
+
+#include "burn_type.H"
+#include "eos.H"
+#include "extern_parameters.H"
+#include "network.H"
+
+using amrex::Real;
+
+struct PrimordialChemTest {
+}; // dummy type to allow compile-type polymorphism via template specialization
+
+// Currently, microphysics uses its own EOS, and this one below is used by hydro. Need to only have one EOS at some point.
+template <> struct quokka::EOS_Traits<PrimordialChemTest> {
+	static constexpr double gamma = 5. / 3.; // default value
+	static constexpr double mean_molecular_weight = quokka::hydrogen_mass_cgs;
+	static constexpr double boltzmann_constant = quokka::boltzmann_constant_cgs;
+};
+
+template <> struct Physics_Traits<PrimordialChemTest> {
+	// cell-centred
+	static constexpr bool is_hydro_enabled = true;
+	static constexpr bool is_chemistry_enabled = false; // in the future, this could point to microphysics, and set to true
+	static constexpr int numPassiveScalars = NumSpec;   // number of chemical species
+	static constexpr bool is_radiation_enabled = false;
+	// face-centred
+	static constexpr bool is_mhd_enabled = false;
+};
+
+// constexpr double Tgas0 = 6000.;	   // K
+// constexpr double rho0 = 0.6 * m_H; // g cm^-3
+
+template <> void RadhydroSimulation<PrimordialChemTest>::setInitialConditionsOnGrid(quokka::grid grid_elem)
+{
+	// set initial conditions
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> dx = grid_elem.dx_;
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> prob_lo = grid_elem.prob_lo_;
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> prob_hi = grid_elem.prob_hi_;
+	const amrex::Box &indexRange = grid_elem.indexRange_;
+	const amrex::Array4<double> &state_cc = grid_elem.array_;
+
+	Real const Lx = (prob_hi[0] - prob_lo[0]);
+	if constexpr (AMREX_SPACEDIM >= 2) {
+		Real const Ly = (prob_hi[1] - prob_lo[1]);
+	}
+	if constexpr (AMREX_SPACEDIM == 3) {
+		Real const Lz = (prob_hi[2] - prob_lo[2]);
+	}
+
+	init_extern_parameters();
+
+	eos_init(small_temp, small_dens);
+	network_init();
```

Comment:
This will actually run for every grid, rather than just once at the beginning of the simulation. Is that what you want here?

Reply:
No, we just want to run it once when starting the simulation. The function it is in is called 'setInitialConditionsOnGrid', but then it also includes the actual microphysics burn. Maybe we should define a separate function for the burn?

Reply:
I think `precalculateInitialConditions` can be used for this purpose.

Reply:
So do I just change the function name from `setInitialConditionsOnGrid` to `precalculateInitialConditions`, or, define an additional function `precalculateInitialConditions` and move the actual burn in there?

Reply:
I think you need to define `precalculateInitialConditions` but only move the initialization there. The actual integration  should only go in `Chemistry.hpp`, not in either of these functions.

Reply:
ah yes, sorry, I got confused. the actual burn is indeed only in `Chemistry.hpp`